node/silkworm/common/object_pool.hpp

```diff
@@ -0,0 +1,55 @@
+/*
+   Copyright 2022 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#ifndef SILKWORM_COMMON_OBJECTPOOL_HPP_
+#define SILKWORM_COMMON_OBJECTPOOL_HPP_
+
+#include <iostream>
+#include <memory>
+#include <stack>
+
+namespace silkworm {
+
+template <class T>
+class ObjectPool {
+  public:
+    using ptr_t = std::unique_ptr<T>;
+
+    ObjectPool() = default;
+    virtual ~ObjectPool() = default;
+
+    void add(std::unique_ptr<T> t) { pool_.push(std::move(t)); }
+
+    ptr_t acquire() {
+        if (!empty()) {
+            ptr_t tmp(pool_.top().release());
+            pool_.pop();
+            return tmp;
+        }
+        return {nullptr};
+    }
+
+    [[nodiscard]] bool empty() const { return pool_.empty(); }
+
+    [[nodiscard]] size_t size() const { return pool_.size(); }
+
+  private:
+    std::stack<std::unique_ptr<T>> pool_;
```

Comment:
I would just use std::vector, where `back()` returns a non-const reference, so you acquire() could return the unique_ptr.

Reply:
But that would mean the vector returns a ref to the unique_ptr.
Here instead the unique_ptr is moved away from it's container and is not guaranteed it returns into the pool

Reply:
you would do: `auto uptr = std::move(pool_.back());`, so grab the unique_ptr from the pool.

Reply:
But why use a vector when there is so nice FIFO stack ?

Reply:
You mean LIFO, right? As far as I know virtually nobody uses std::stack. Why deal with a different class when `std::vector` is a stack when you `push_back` and `pop_back`, and has a better api.

Reply:
Yeah LIFO. Sorry.
Well honestly I think I don't need to bother with contiguous memory. The underlying container is a deque which provides better perf imho especially when the pool gets large.

Reply:
> provides better perf imho especially when the pool gets large

It would have to grow quite large to make a difference. I guess it is a matter of preference at this point, and there is nothing wrong with your version :-).

Reply:
Well i thought object pool to be as abstract as possible so you never know how many elements. Anyhow keep pushing your comments: always extremely appreciated